proposals/NNNN-resource-arrays.md

```diff
@@ -0,0 +1,258 @@
+
+A group of resources can be declared as one-dimensional array of fixed size:
+
+```
+RWBuffer<float> A[4] : register(u10);
+```
+
+Each element of the array corresponds to a single resource that is initialized
+by an individual `@dx.op.createHandleFromBinding` DXIL intrinsic call. The fact
+that this resource is part of an array is encoded in the `dx.types.ResBind`
+argument of the call. The `dx.types.ResBind` structure contains binding
+information for the whole array - the lower bound and upper bound values for the
+register slots and a virtual register space. Third argument of
+`@dx.op.createHandleFromBinding` is an index of the individual resource in the
+descriptor range from the range `<lower bound, upper bound>` (inclusive).
+
+For example for the resource `A` declared above that is bound to register `u10`,
+the DXIL call to initialize resource handle for `A[2]` will have lower bound
+value `10`, upper bound `13`, space `0` and array index of `2`:
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 10, i32 13, i32 0, i8 1 }, i32 2, i1 false),
+```
+
+Initialization of other resources in the same array would differ only by the
+index. The values for upper and lower bound and register space will stay the
+same.
+
+https://godbolt.org/z/ff3bnKxov
+
+#### Fixed-size Multi-dimensional Array
+
+Groups of resources can also be declared as multi-dimensional arrays of fixed
+size
+
+```
+RWBuffer<float> B[4][4] : register(u2);
+RWBuffer<float> C[2][2][5] : register(u10, space1);
+```
+
+When initializing individual resources of multi-dimensional array the arguments
+of `@dx.op.createHandleFromBinding` look like as if the array was flattened to a
+single dimension.
+
+That means the array `B[4][4]` is treated as if it was declared as `B[16]`, and
+`C[2][2][5]` as if it was `C[20]`.
+
+For example a call to initialize `C[1][0][3][0]` from the array declared above
+would have lower bound value `10`, upper bound `29`, space `1`, and array index
+`23` = `10 + 1*2*5 + 0*5 + 3`:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 10, i32 29, i32 1, i8 1 }, i32 23, i1 false), 
+```
+
+https://godbolt.org/z/8MMWheeGc
+
+#### Arrays of Unbounded Size
+
+Groups of resource arrays can be declared with undefined size. These are called
+_unbounded arrays_:
+```
+RWBuffer<float> D[] : register(u5);
+```
+For unbounded arrays the `@dx.op.createHandleFromBinding` calls to initialize
+individual resource handles will have upper bound set to `-1`. For example an
+initialization of handle `D[100]` from the `D` array declared above looks like
+this:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 105, i1 false)
+```
+https://godbolt.org/z/hx99s5dzP
+
+#### Unbounded-size Arrays and Multiple Dimensions
+
+For multi-dimensional arrays of resources only the first dimension can be
+unbounded:
```

Comment:
This is an oddity of the C array grammar. The dimension of a base array can be unbounded, but you can't have an array where the element size is unknown, so the element must be a complete type (hence the weird diagnostic). For your example below you're declaring an unbounded array of `RWBuffer<float>[4]` elements. It makes sense when you think about how multi-dimensional array indexing works, but isn't directly intuitive (at least it isn't to me).